Canon CR2: white balance

LebedevRI opened the pull request:
This was developed against DT copy of rawspeed, that already contains all those yet unmerged changes by @pedrocr 
Specifically, it needs macrosed like get2BE() that are part of PR #73 
Until that PR is merged, this one _should_ not be merged.

With this, RawSpeed can read WB info from ALL CR2 files (at least at this moment, all those cr2 sample files from rawsamples.ch)

LebedevRI committed:
Cr2Decoder: WB: support most cameras

This adds supports for getting WB from all those cameras listed in switch statement.

I have originally taken that switch from rawstudio, and re-checked against
all CR2 files that are currently available from rawsamples.ch

Known cameras with CR2 that are not supported yet are:
1) All  Canon PowerShot

LebedevRI committed:
Cr2Decoder: WB: support Canon PowerShot cameras

This supports all those cameras listed in switch statement.

Only one (known) camera with CR2 that is not supported yet:
1) Canon PowerShot G9

klauspost commented on RawSpeed/Cr2Decoder.cpp:
```diff
@@ -190,6 +191,104 @@ void Cr2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
+        offset += 142;
+      break;
+      default:
+        std::cerr << "CR2 Decoder: CanonColorData has unsupported count of values: %d" << color_data->count << std::endl;
```
Please don't print to stderr. Use writeLog(DEBUG_PRIO_INFO, "CR2 Decoder: CanonColorData has unsupported count of values: %d", color_data->count);

klauspost commented on RawSpeed/Cr2Decoder.cpp:
```diff
@@ -190,6 +191,104 @@ void Cr2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
+    // this entry is a big table, and different cameras store used WB in
+    // different parts, so find the offset
+    int offset = 0;
+    switch (color_data->count)
```
Please use this value as a _default_ value, add a hint that allow us to override the value like this

`<Hint name="wb_offset" value="72"/>`

That way, we can add support for new cameras without modifying code.

klauspost commented on RawSpeed/Cr2Decoder.cpp:
```diff
@@ -190,6 +191,104 @@ void Cr2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
+      float cam_mul[4];
+      for(int c = 0; c < 4; c++)
+      {
+        cam_mul[c] = (float) data[offset/2 + c];
```
data[offset/2 + c] needs to be bound-schecked against the tiff entry, otherwise we can read junk/crash.

klauspost commented:
Looks good. I had a quick look, and added a few notes. #73 should be merged soon, so this comes at a good time!

LebedevRI committed:
Cr2Decoder: WB: do not print to stderr

LebedevRI committed:
Cr2Decoder: WB: convert switch to <Hint/>

That way, we can add support for new cameras without modifying code.

LebedevRI committed:
Cr2Decoder: WB: support Canon PowerShot G9

This concludes WB support for CR2 cameras.

All known cameras with CR2 are now supported.

getIntArray() doesn't accept TIFF_BYTE, has to use get4LE()

FIXME: do we really need to check that this is Canon PowerShot G9?

pedrocr commented:
I've fixed the 1D/1Ds support including WB for them so after you merge this one I'll update the PR

klauspost commented:
Cool - just fixing a leak in CRW. After I have verified it, I will look at this.
